File: tools/eos/saas/domain/employee.py
# -*- coding: utf-8 -*-
"""
╔══════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗
║ WILSY OS – SOVEREIGN EMPLOYEE DOMAIN MODEL (PYTHON)                                                            ║
╠══════════════════════════════════════════════════════════════════════════════════════════════════════════════════╣
║ FILE:           tools/eos/saas/domain/employee.py                                                              ║
║ VERSION:        v1.0.0-INSTITUTIONAL                                                                           ║
║ AUTHORITY:      Wilsy OS Core Governance                                                                       ║
║ EPITOME:        Python dataclass mirroring Node Employee model, for employee search and management.            ║
║ CLASSIFICATION: Production Artifact                                                                             ║
╠══════════════════════════════════════════════════════════════════════════════════════════════════════════════════╣
║ 🔧 CHANGE LOG:                                                                                                  ║
║   2026-08-20 v1.0.0-INSTITUTIONAL – Initial creation.                                                          ║
╠══════════════════════════════════════════════════════════════════════════════════════════════════════════════════╣
║ COMPLIANCE:    POPIA §19 │ GDPR §32 │ SOC2 §CC7.2 │ ISO 27001                                                  ║
║ IDENTITY:      employeeId (string, e.g., WIL-001)                                                              ║
║ INTEGRATION:   Used by employee_registry.py and employee_router.py.                                            ║
╚══════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

# ─── Helper ──────────────────────────────────────────────────────────────────

def parse_datetime(val: Any) -> Optional[datetime]:
    if isinstance(val, datetime):
        return val
    if isinstance(val, str):
        try:
            return datetime.fromisoformat(val.replace('Z', '+00:00'))
        except ValueError:
            return None
    return None

# ─── Nested dataclasses ────────────────────────────────────────────────────

@dataclass
class LegalName:
    firstName: str
    lastName: str
    middleName: Optional[str] = None
    suffix: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "firstName": self.firstName,
            "lastName": self.lastName,
            "middleName": self.middleName,
            "suffix": self.suffix,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "LegalName":
        return cls(
            firstName=data["firstName"],
            lastName=data["lastName"],
            middleName=data.get("middleName"),
            suffix=data.get("suffix"),
        )

@dataclass
class ContactInfo:
    personalEmail: Optional[str] = None
    workEmail: Optional[str] = None
    personalPhone: Optional[str] = None
    workPhone: Optional[str] = None
    emergencyContact: Optional[Dict[str, Any]] = None  # name, relationship, phone

    def to_dict(self) -> Dict[str, Any]:
        return {
            "personalEmail": self.personalEmail,
            "workEmail": self.workEmail,
            "personalPhone": self.personalPhone,
            "workPhone": self.workPhone,
            "emergencyContact": self.emergencyContact,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "ContactInfo":
        return cls(
            personalEmail=data.get("personalEmail"),
            workEmail=data.get("workEmail"),
            personalPhone=data.get("personalPhone"),
            workPhone=data.get("workPhone"),
            emergencyContact=data.get("emergencyContact"),
        )

@dataclass
class Address:
    street: str
    city: str
    state: str
    postalCode: str
    country: str

    def to_dict(self) -> Dict[str, Any]:
        return {
            "street": self.street,
            "city": self.city,
            "state": self.state,
            "postalCode": self.postalCode,
            "country": self.country,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "Address":
        return cls(
            street=data["street"],
            city=data["city"],
            state=data["state"],
            postalCode=data["postalCode"],
            country=data["country"],
        )

@dataclass
class EmployeeEntity:
    """
    Sovereign employee entity – mirrors Node Employee document.
    """
    # Identity
    employeeId: str
    legalName: LegalName
    displayName: str
    # Contact
    contact: ContactInfo
    # Addresses
    physicalAddress: Address
    postalAddress: Address
    # Employment
    jobTitle: str
    department: str
    employmentType: str
    status: str  # ACTIVE, INACTIVE, etc.
    hireDate: datetime
    workLocation: str
    # System
    tenantId: str
    isActive: bool = True
    # Optional fields
    externalId: Optional[str] = None
    preferredName: Optional[str] = None
    photograph: Optional[str] = None
    dateOfBirth: Optional[datetime] = None
    placeOfBirth: Optional[Dict[str, str]] = None
    gender: Optional[str] = None
    maritalStatus: Optional[str] = None
    nationality: Optional[str] = None
    language: Optional[str] = None
    religion: Optional[str] = None
    ethnicity: Optional[str] = None
    managerId: Optional[str] = None
    seniorityDate: Optional[datetime] = None
    terminationDate: Optional[datetime] = None
    basicSalary: float = 0.0
    salaryCurrency: str = "ZAR"
    payFrequency: str = "MONTHLY"
    costCentre: Optional[str] = None
    onboardingStatus: str = "COMPLETED"
    probationEndDate: Optional[datetime] = None
    leaveBalance: Dict[str, float] = field(default_factory=lambda: {"annual": 20, "sick": 10, "familyResponsibility": 3})
    skills: List[str] = field(default_factory=list)
    qualifications: List[str] = field(default_factory=list)
    backgroundCheck: Optional[str] = None
    criminalRecord: Optional[str] = None
    identification: Dict[str, Any] = field(default_factory=dict)  # nationalId, passport, taxId, etc.
    financial: Dict[str, Any] = field(default_factory=dict)  # bankAccount
    compliance: Dict[str, Any] = field(default_factory=dict)
    hr: Dict[str, Any] = field(default_factory=dict)  # joinedDate, onboardingStatus, etc.
    metadata: Dict[str, Any] = field(default_factory=dict)
    deletedAt: Optional[datetime] = None
    createdAt: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    updatedAt: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    def to_dict(self) -> Dict[str, Any]:
        """Serialise to dictionary (camelCase for frontend compatibility)."""
        return {
            "employeeId": self.employeeId,
            "externalId": self.externalId,
            "legalName": self.legalName.to_dict(),
            "preferredName": self.preferredName,
            "displayName": self.displayName,
            "photograph": self.photograph,
            "dateOfBirth": self.dateOfBirth.isoformat() if self.dateOfBirth else None,
            "placeOfBirth": self.placeOfBirth,
            "gender": self.gender,
            "maritalStatus": self.maritalStatus,
            "nationality": self.nationality,
            "language": self.language,
            "religion": self.religion,
            "ethnicity": self.ethnicity,
            "identification": self.identification,
            "contact": self.contact.to_dict(),
            "address": {
                "physical": self.physicalAddress.to_dict(),
                "postal": self.postalAddress.to_dict(),
            },
            "employment": {
                "jobTitle": self.jobTitle,
                "department": self.department,
                "managerId": self.managerId,
                "employmentType": self.employmentType,
                "status": self.status,
                "hireDate": self.hireDate.isoformat() if self.hireDate else None,
                "seniorityDate": self.seniorityDate.isoformat() if self.seniorityDate else None,
                "terminationDate": self.terminationDate.isoformat() if self.terminationDate else None,
                "workLocation": self.workLocation,
            },
            "financial": {
                "bankAccount": self.financial.get("bankAccount", {}),
                "basicSalary": self.basicSalary,
                "salaryCurrency": self.salaryCurrency,
                "payFrequency": self.payFrequency,
                "costCentre": self.costCentre,
            },
            "compliance": self.compliance,
            "hr": {
                "joinedDate": self.hireDate.isoformat() if self.hireDate else None,
                "onboardingStatus": self.onboardingStatus,
                "probationEndDate": self.probationEndDate.isoformat() if self.probationEndDate else None,
                "leaveBalance": self.leaveBalance,
                "skills": self.skills,
                "qualifications": self.qualifications,
            },
            "tenantId": self.tenantId,
            "isActive": self.isActive,
            "deletedAt": self.deletedAt.isoformat() if self.deletedAt else None,
            "metadata": self.metadata,
            "createdAt": self.createdAt.isoformat(),
            "updatedAt": self.updatedAt.isoformat(),
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "EmployeeEntity":
        """Deserialise from dictionary (MongoDB document)."""
        legal_name = LegalName.from_dict(data["legalName"])
        contact = ContactInfo.from_dict(data.get("contact", {}))
        addr_physical = Address.from_dict(data.get("address", {}).get("physical", {}))
        addr_postal = Address.from_dict(data.get("address", {}).get("postal", {}))
        # Flatten employment fields
        emp = data.get("employment", {})
        fin = data.get("financial", {})
        hr = data.get("hr", {})
        return cls(
            employeeId=data["employeeId"],
            legalName=legal_name,
            displayName=data["displayName"],
            contact=contact,
            physicalAddress=addr_physical,
            postalAddress=addr_postal,
            jobTitle=emp.get("jobTitle", ""),
            department=emp.get("department", ""),
            employmentType=emp.get("employmentType", "PERMANENT"),
            status=emp.get("status", "ACTIVE"),
            hireDate=parse_datetime(emp.get("hireDate")) or datetime.now(timezone.utc),
            workLocation=emp.get("workLocation", "Remote"),
            tenantId=data["tenantId"],
            isActive=data.get("isActive", True),
            externalId=data.get("externalId"),
            preferredName=data.get("preferredName"),
            photograph=data.get("photograph"),
            dateOfBirth=parse_datetime(data.get("dateOfBirth")),
            placeOfBirth=data.get("placeOfBirth"),
            gender=data.get("gender"),
            maritalStatus=data.get("maritalStatus"),
            nationality=data.get("nationality"),
            language=data.get("language"),
            religion=data.get("religion"),
            ethnicity=data.get("ethnicity"),
            managerId=emp.get("managerId"),
            seniorityDate=parse_datetime(emp.get("seniorityDate")),
            terminationDate=parse_datetime(emp.get("terminationDate")),
            basicSalary=fin.get("basicSalary", 0.0),
            salaryCurrency=fin.get("salaryCurrency", "ZAR"),
            payFrequency=fin.get("payFrequency", "MONTHLY"),
            costCentre=fin.get("costCentre"),
            onboardingStatus=hr.get("onboardingStatus", "COMPLETED"),
            probationEndDate=parse_datetime(hr.get("probationEndDate")),
            leaveBalance=hr.get("leaveBalance", {"annual": 20, "sick": 10, "familyResponsibility": 3}),
            skills=hr.get("skills", []),
            qualifications=hr.get("qualifications", []),
            backgroundCheck=data.get("backgroundCheck"),
            criminalRecord=data.get("criminalRecord"),
            identification=data.get("identification", {}),
            financial=data.get("financial", {}),
            compliance=data.get("compliance", {}),
            hr=data.get("hr", {}),
            metadata=data.get("metadata", {}),
            deletedAt=parse_datetime(data.get("deletedAt")),
            createdAt=parse_datetime(data.get("createdAt")) or datetime.now(timezone.utc),
            updatedAt=parse_datetime(data.get("updatedAt")) or datetime.now(timezone.utc),
        )

"""
════════════════════════════════════════════════════════════════════════════════
INSTITUTIONAL CERTIFICATION SEAL — WILSY OS EMPLOYEE DOMAIN
════════════════════════════════════════════════════════════════════════════════
Status:          CERTIFIED PRODUCTION ARTIFACT
Version:         v1.0.0-INSTITUTIONAL
Compliance:      POPIA §19 │ GDPR §32 │ SOC2 §CC7.2 │ ISO 27001
Pending Work:    None – ready for registry and router.
════════════════════════════════════════════════════════════════════════════════
"""

File: server/models/Employee.js
/**
 * ╔════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗
 * ║ WILSY OS – SOVEREIGN EMPLOYEE MODEL [V3.0.4-SYNC-HOOK]                                                                               ║
 * ║ [HUMAN RESOURCES FABRIC | CRM | BILLING | PAYROLL | COMPLIANCE]                                                                       ║
 * ╠════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╣
 * ║ VERSION: 3.0.4-SYNC-HOOK | PRODUCTION READY                                                                                           ║
 * ║ EPITOME: Mongoose 7+/8-safe pre-save (no next callback). Uniqueness via DB index.                                                   ║
 * ║ ABSOLUTE PATH: /Users/wilsonkhanyezi/legal-doc-system/server/models/Employee.js                                                       ║
 * ╠════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╣
 * ║ 🔧 FIX (v3.0.4):                                                                                                                       ║
 * ║   1. Removed next() from pre('save') — Mongoose no longer injects next; calling it → "next is not a function".                      ║
 * ║   2. Sync hook only sets displayName when empty; safe optional chaining on legalName.                                                 ║
 * ║   3. Unique indexes remain the source of truth for employeeId / workEmail.                                                            ║
 * ║   4. All fields preserved from V3.0.3.                                                                                                 ║
 * ╚════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝
 */

import mongoose from 'mongoose';

const { Schema } = mongoose;

// ─── ENUMS ────────────────────────────────────────────────────────────────────

const EMPLOYMENT_TYPES = ['PERMANENT', 'CONTRACT', 'INTERN', 'TRAINEE', 'CONSULTANT', 'PROBATION'];
const EMPLOYEE_STATUSES = ['ACTIVE', 'ON_LEAVE', 'SUSPENDED', 'TERMINATED', 'RETIRED', 'INACTIVE'];
const GENDER_OPTIONS = ['MALE', 'FEMALE', 'NON_BINARY', 'PREFER_NOT_TO_SAY', 'OTHER'];
const MARITAL_STATUSES = ['SINGLE', 'MARRIED', 'DIVORCED', 'WIDOWED', 'SEPARATED', 'PREFER_NOT_TO_SAY'];
const PAY_FREQUENCIES = ['MONTHLY', 'BI_WEEKLY', 'WEEKLY', 'DAILY', 'ANNUAL'];
const ONBOARDING_STATUSES = ['NOT_STARTED', 'PENDING', 'IN_PROGRESS', 'COMPLETED', 'FAILED'];
const IDENTIFICATION_TYPES = ['NATIONAL_ID', 'PASSPORT', 'DRIVERS_LICENSE', 'WORK_PERMIT', 'VISA', 'OTHER'];
const WORK_PERMIT_TYPES = ['GENERAL', 'CRITICAL_SKILLS', 'INTRA_COMPANY', 'STUDENT', 'OTHER'];
const VISA_TYPES = ['TOURIST', 'BUSINESS', 'WORK', 'STUDENT', 'RESIDENCE', 'DIPLOMATIC', 'OTHER'];

// ─── SUB‑SCHEMAS ─────────────────────────────────────────────────────────────

const NationalIdSchema = new Schema({
  number: { type: String, required: true, trim: true },
  country: { type: String, required: true, trim: true, default: 'ZA' },
  type: { type: String, enum: IDENTIFICATION_TYPES, default: 'NATIONAL_ID' },
  issueDate: { type: Date, default: null },
  expiryDate: { type: Date, default: null },
  issuingAuthority: { type: String, trim: true, default: '' },
}, { _id: false });

const PassportSchema = new Schema({
  number: { type: String, required: true, trim: true },
  country: { type: String, required: true, trim: true, default: 'ZA' },
  issueDate: { type: Date, default: null },
  expiryDate: { type: Date, default: null },
  issuingAuthority: { type: String, trim: true, default: '' },
}, { _id: false });

const WorkPermitSchema = new Schema({
  number: { type: String, required: true, trim: true },
  type: { type: String, enum: WORK_PERMIT_TYPES, default: 'GENERAL' },
  issuingCountry: { type: String, required: true, trim: true, default: 'ZA' },
  issueDate: { type: Date, default: null },
  expiryDate: { type: Date, default: null },
}, { _id: false });

const VisaSchema = new Schema({
  number: { type: String, required: true, trim: true },
  type: { type: String, enum: VISA_TYPES, default: 'WORK' },
  issuingCountry: { type: String, required: true, trim: true, default: 'ZA' },
  issueDate: { type: Date, default: null },
  expiryDate: { type: Date, default: null },
}, { _id: false });

const DriversLicenseSchema = new Schema({
  number: { type: String, required: true, trim: true },
  country: { type: String, required: true, trim: true, default: 'ZA' },
  class: { type: String, trim: true, default: '' },
  issueDate: { type: Date, default: null },
  expiryDate: { type: Date, default: null },
}, { _id: false });

// ─── MAIN SCHEMA ─────────────────────────────────────────────────────────────

const EmployeeSchema = new Schema(
  {
    // ─── 1. IDENTITY ──────────────────────────────────────────────────────────
    employeeId: {
      type: String,
      required: [true, 'Employee ID is required'],
      unique: true,
      trim: true,
      index: true,
    },
    externalId: {
      type: String,
      trim: true,
      sparse: true,
      index: true,
    },
    legalName: {
      firstName: { type: String, required: true, trim: true, index: true },
      middleName: { type: String, trim: true, default: '' },
      lastName: { type: String, required: true, trim: true, index: true },
      suffix: { type: String, trim: true, default: '' },
    },
    preferredName: {
      type: String,
      trim: true,
      default: '',
    },
    displayName: {
      type: String,
      trim: true,
      default: '',
    },
    photograph: {
      type: String,
      trim: true,
      default: '',
    },

    // ─── 2. PERSONAL ──────────────────────────────────────────────────────────
    dateOfBirth: {
      type: Date,
      required: true,
    },
    placeOfBirth: {
      city: { type: String, trim: true, default: '' },
      country: { type: String, trim: true, default: '' },
    },
    gender: {
      type: String,
      enum: GENDER_OPTIONS,
      default: 'PREFER_NOT_TO_SAY',
    },
    maritalStatus: {
      type: String,
      enum: MARITAL_STATUSES,
      default: 'PREFER_NOT_TO_SAY',
    },
    nationality: {
      type: String,
      trim: true,
      default: '',
      index: true,
    },
    language: {
      type: String,
      trim: true,
      default: 'en',
    },
    religion: {
      type: String,
      trim: true,
      default: '',
    },
    ethnicity: {
      type: String,
      trim: true,
      default: '',
    },

    // ─── 3. IDENTITY DOCUMENTATION ──────────────────────────────────────────
    identification: {
      nationalId: { type: NationalIdSchema, default: null },
      passport: { type: PassportSchema, default: null },
      taxId: {
        number: { type: String, trim: true, default: '' },
        country: { type: String, trim: true, default: 'ZA' },
      },
      workPermit: { type: WorkPermitSchema, default: null },
      visa: { type: VisaSchema, default: null },
      driversLicense: { type: DriversLicenseSchema, default: null },
      socialSecurity: {
        number: { type: String, trim: true, default: '' },
        country: { type: String, trim: true, default: 'ZA' },
      },
      additional: {
        type: [{
          type: { type: String, enum: IDENTIFICATION_TYPES, default: 'OTHER' },
          number: { type: String, trim: true },
          country: { type: String, trim: true, default: 'ZA' },
          issueDate: { type: Date, default: null },
          expiryDate: { type: Date, default: null },
          issuingAuthority: { type: String, trim: true, default: '' },
        }],
        default: [],
      },
    },

    // ─── 4. CONTACT ───────────────────────────────────────────────────────────
    contact: {
      personalEmail: {
        type: String,
        trim: true,
        lowercase: true,
        match: [/^\S+@\S+\.\S+$/, 'Please provide a valid email address'],
      },
      workEmail: {
        type: String,
        required: true,
        unique: true,
        trim: true,
        lowercase: true,
        match: [/^\S+@\S+\.\S+$/, 'Please provide a valid email address'],
        index: true,
      },
      personalPhone: { type: String, trim: true, default: '' },
      workPhone: { type: String, trim: true, default: '' },
      emergencyContact: {
        name: { type: String, trim: true, default: '' },
        relationship: { type: String, trim: true, default: '' },
        phone: { type: String, trim: true, default: '' },
      },
    },

    // ─── 5. ADDRESS ───────────────────────────────────────────────────────────
    address: {
      physical: {
        street: { type: String, trim: true, default: '' },
        city: { type: String, trim: true, default: '' },
        state: { type: String, trim: true, default: '' },
        postalCode: { type: String, trim: true, default: '' },
        country: { type: String, trim: true, default: 'ZA' },
      },
      postal: {
        street: { type: String, trim: true, default: '' },
        city: { type: String, trim: true, default: '' },
        state: { type: String, trim: true, default: '' },
        postalCode: { type: String, trim: true, default: '' },
        country: { type: String, trim: true, default: 'ZA' },
      },
    },

    // ─── 6. EMPLOYMENT ────────────────────────────────────────────────────────
    employment: {
      jobTitle: { type: String, required: true, trim: true, index: true },
      department: { type: String, required: true, trim: true, index: true },
      managerId: { type: Schema.Types.ObjectId, ref: 'Employee', default: null, index: true },
      employmentType: { type: String, enum: EMPLOYMENT_TYPES, required: true, default: 'PERMANENT' },
      status: { type: String, enum: EMPLOYEE_STATUSES, required: true, default: 'ACTIVE', index: true },
      hireDate: { type: Date, required: true },
      seniorityDate: { type: Date, default: null },
      terminationDate: { type: Date, default: null },
      workLocation: { type: String, trim: true, default: '' },
    },

    // ─── 7. FINANCIAL ────────────────────────────────────────────────────────
    financial: {
      bankAccount: {
        accountName: { type: String, trim: true, default: '' },
        accountNumber: { type: String, trim: true, default: '' },
        branchCode: { type: String, trim: true, default: '' },
        bankName: { type: String, trim: true, default: '' },
      },
      basicSalary: { type: Number, default: 0, min: 0 },
      salaryCurrency: { type: String, default: 'ZAR', uppercase: true, trim: true },
      payFrequency: { type: String, enum: PAY_FREQUENCIES, default: 'MONTHLY' },
      costCentre: { type: String, trim: true, default: '' },
    },

    // ─── 8. COMPLIANCE ────────────────────────────────────────────────────────
    compliance: {
      backgroundCheck: { type: String, trim: true, default: '' },
      backgroundCheckDate: { type: Date, default: null },
      criminalRecord: { type: String, trim: true, default: '' },
    },

    // ─── 9. HR / ADMIN ────────────────────────────────────────────────────────
    hr: {
      joinedDate: { type: Date, default: Date.now },
      onboardingStatus: { type: String, enum: ONBOARDING_STATUSES, default: 'NOT_STARTED' },
      probationEndDate: { type: Date, default: null },
      leaveBalance: {
        annual: { type: Number, default: 0, min: 0 },
        sick: { type: Number, default: 0, min: 0 },
        familyResponsibility: { type: Number, default: 0, min: 0 },
      },
      skills: { type: [String], default: [], index: true },
      qualifications: { type: [String], default: [] },
    },

    // ─── 10. SYSTEM ────────────────────────────────────────────────────────────
    tenantId: {
      type: String,
      required: [true, 'tenantId is required'],
      trim: true,
      index: true,
      default: 'GLOBAL_ROOT',
    },
    isActive: { type: Boolean, default: true, index: true },
    deletedAt: { type: Date, default: null, index: true },
    metadata: { type: Schema.Types.Mixed, default: {} },
  },
  {
    timestamps: true,
    toJSON: {
      transform: (doc, ret) => {
        ret.id = ret._id;
        delete ret._id;
        delete ret.__v;
        return ret;
      },
    },
    toObject: {
      transform: (doc, ret) => {
        ret.id = ret._id;
        delete ret._id;
        delete ret.__v;
        return ret;
      },
    },
  }
);

// ─── INDEXES ──────────────────────────────────────────────────────────────────

EmployeeSchema.index({ tenantId: 1, isActive: 1, 'employment.status': 1 });
EmployeeSchema.index({ tenantId: 1, 'employment.department': 1 });
EmployeeSchema.index({ tenantId: 1, 'employment.jobTitle': 1 });
EmployeeSchema.index({ 'employment.managerId': 1, tenantId: 1 });

EmployeeSchema.index({ 'identification.nationalId.number': 1 });
EmployeeSchema.index({ 'identification.passport.number': 1 });
EmployeeSchema.index({ 'identification.taxId.number': 1 });

EmployeeSchema.index(
  {
    employeeId: 'text',
    'legalName.firstName': 'text',
    'legalName.lastName': 'text',
    displayName: 'text',
    'contact.workEmail': 'text',
    'hr.skills': 'text',
    'identification.nationalId.number': 'text',
    'identification.passport.number': 'text',
  },
  {
    name: 'employee_text_search',
    weights: {
      employeeId: 10,
      'legalName.firstName': 8,
      'legalName.lastName': 8,
      displayName: 6,
      'contact.workEmail': 4,
      'hr.skills': 2,
      'identification.nationalId.number': 5,
      'identification.passport.number': 5,
    },
  }
);

// ─── HOOKS (Mongoose 7+/8 — do NOT use next) ─────────────────────────────────

EmployeeSchema.pre('save', function setDisplayName() {
  if (!this.displayName || String(this.displayName).trim() === '') {
    const parts = [
      this.legalName?.firstName,
      this.legalName?.lastName,
    ].filter(Boolean);
    this.displayName = parts.join(' ');
  }
});

// ─── STATIC METHODS ──────────────────────────────────────────────────────────

EmployeeSchema.statics.findActiveByTenant = function (tenantId, filters = {}, limit = 100) {
  const query = { tenantId, isActive: true, deletedAt: null };
  if (filters.department) query['employment.department'] = filters.department;
  if (filters.jobTitle) query['employment.jobTitle'] = filters.jobTitle;
  if (filters.status) query['employment.status'] = filters.status;
  return this.find(query).sort({ 'legalName.lastName': 1, 'legalName.firstName': 1 }).limit(limit).lean();
};

EmployeeSchema.statics.findDirectReports = function (managerId, tenantId) {
  return this.find({
    'employment.managerId': managerId,
    tenantId,
    isActive: true,
    deletedAt: null,
  }).sort({ 'legalName.lastName': 1 }).lean();
};

EmployeeSchema.statics.searchEmployees = function (query, tenantId, options = {}) {
  const limit = Math.min(options.limit || 20, 100);
  const offset = options.offset || 0;
  const filter = { tenantId, isActive: true, deletedAt: null };
  let searchQuery = {};
  if (query && query.trim().length >= 2) {
    searchQuery = { $text: { $search: query.trim() } };
  }
  const finalFilter = { ...filter, ...searchQuery };
  return Promise.all([
    this.find(finalFilter)
      .sort(query && query.trim().length >= 2 ? { score: { $meta: 'textScore' } } : { 'legalName.lastName': 1 })
      .skip(offset)
      .limit(limit)
      .lean(),
    this.countDocuments(finalFilter),
  ]).then(([items, total]) => ({ items, total }));
};

// ─── MODEL EXPORT ──────────────────────────────────────────────────────────

const Employee = mongoose.models.Employee || mongoose.model('Employee', EmployeeSchema);
export default Employee;

/**
 * ═══════════════════════════════════════════════════════════════════════════════
 * 🏛️ INSTITUTIONAL CERTIFICATION SEAL — Employee Model V3.0.4-SYNC-HOOK
 * ═══════════════════════════════════════════════════════════════════════════════
 * Status:          CERTIFIED PRODUCTION ARTIFACT
 * Fix:             pre('save') no longer calls next() — fixes TypeError under Mongoose 7/8
 * Compliance:      POPIA §19 · GDPR §32 · SOC2 §CC7.2 · ISO 27001 · ECT Act §15
 * ═══════════════════════════════════════════════════════════════════════════════
 */
